### PR TITLE
Project reactor

### PR DESCRIPTION
**Pull request by Wectro20**

### CODE REVIEW

**Comment by ajax-gnatiuk-s** [comment]
@tetty-pa please review

**Comment by Wectro20** [comment]
I have problems with saving csv file, trying to fix it as soon as i can

**Comment by Wectro20** [comment]
@tetty-pa review pls

File: src/main/kotlin/com/ajax/cryptocurrency/service/CryptocurrencyService.kt [hunk 1]
```diff
@@ -2,48 +2,67 @@ package com.ajax.cryptocurrency.service
...
+                    cryptoInfoList.forEach { (name, minPrice, maxPrice) ->
+                        writer.append("$name,$minPrice,$maxPrice\n")
+                    }
+                }
+                .collectList()
+                .doOnSuccess {
+                    writer.close()
+                }
+                .thenReturn(file)
+        }
```
Comment (ajax-gnatiuk-s) [inline src/main/kotlin/com/ajax/cryptocurrency/service/CryptocurrencyService.kt:66]: Could be simplified
```kotlin
        return Flux.fromIterable(cryptocurrencies)
            .flatMap { name ->
                Mono.zip(
                    cryptocurrencyRepository.findMinMaxByName(name, 1),
                    cryptocurrencyRepository.findMinMaxByName(name, -1)
                ).map {
                    "$name,${it.t1},${it.t1}\n"
                }
            }
            .collectList()
            .map { list ->
                file.bufferedWriter().use { writer ->
                    writer.append("Cryptocurrency Name,Min Price,Max Price\n")
                    list.forEach { writer.append(it) }
                }
                file
            }
```

**Review by ajax-gnatiuk-s — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/ajax/cryptocurrency/controller/CryptocurrencyController.kt [hunk 2]
```diff
@@ -20,38 +22,41 @@ import org.springframework.web.bind.annotation.RestController
...
 
     @GetMapping("/minprice")
-    fun getMinCryptocurrencyPrice(@RequestParam name: String): ResponseEntity<Cryptocurrency> =
+    fun getMinCryptocurrencyPrice(@RequestParam name: String): ResponseEntity<Mono<Cryptocurrency>> =
         ResponseEntity.ok(cryptocurrencyService.findMinMaxPriceByCryptocurrencyName(name, 1))
 
 
     @GetMapping("/maxprice")
-    fun getMaxCryptocurrencyPrice(@RequestParam() name: String): ResponseEntity<Cryptocurrency> =
+    fun getMaxCryptocurrencyPrice(@RequestParam() name: String): ResponseEntity<Mono<Cryptocurrency>> =
```
[the remark this answers is not in the document]
Comment (Wectro20) [reply src/main/kotlin/com/ajax/cryptocurrency/controller/CryptocurrencyController.kt:34]: done

File: src/test/kotlin/com/ajax/cryptocurrency/natscontroller/NatsCryptocurrencyGetMinPriceControllerTest.kt [hunk 3]
```diff
@@ -75,7 +73,7 @@ class NatsCryptocurrencyGetMinPriceControllerTest {
         assertEquals(crypto.price, cryptoFromResponse.price)
         assertEquals(crypto.createdTime, cryptoFromResponse.createdTime)
 
-        verify { cryptocurrencyService.findMinMaxPriceByCryptocurrencyName(cryptoName, 1) }
+        verify { cryptocurrencyService.findMinMaxPriceByCryptocurrencyName(cryptoName, 1).block() }
```
[the remark this answers is not in the document]
Comment (Wectro20) [reply src/test/kotlin/com/ajax/cryptocurrency/natscontroller/NatsCryptocurrencyGetMinPriceControllerTest.kt:76]: done

File: src/main/kotlin/com/ajax/cryptocurrency/service/CryptocurrencyService.kt (hunk 1 above)
Comment (Wectro20) [reply src/main/kotlin/com/ajax/cryptocurrency/service/CryptocurrencyService.kt:66]: thx for hint

File: src/main/kotlin/com/ajax/cryptocurrency/nats/NatsCryptocurrencyGetMinPriceController.kt [hunk 4]
```diff
@@ -21,7 +21,7 @@ class NatsCryptocurrencyGetMinPriceController(
     override val parser: Parser<CryptocurrencyRequest> = CryptocurrencyRequest.parser()
 
     override fun handler(request: CryptocurrencyRequest): CryptocurrencyResponse {
-        val cryptocurrency = cryptocurrencyService.findMinMaxPriceByCryptocurrencyName(request.name.name, 1)
+        val cryptocurrency = cryptocurrencyService.findMinMaxPriceByCryptocurrencyName(request.name.name, 1).block()!!
```
[the remark this answers is not in the document]
Comment (Wectro20) [reply src/main/kotlin/com/ajax/cryptocurrency/nats/NatsCryptocurrencyGetMinPriceController.kt:24]: I'll change it to your version, I like it better

[the remark this answers is not in the document]
Comment (Wectro20) [reply src/main/kotlin/com/ajax/cryptocurrency/nats/NatsCryptocurrencyGetMinPriceController.kt:24]: done

File: src/test/kotlin/com/ajax/cryptocurrency/natscontroller/NatsCryptocurrencyGetAllControllerTest.kt [hunk 5]
```diff
@@ -50,7 +50,7 @@ class NatsCryptocurrencyGetAllControllerTest {
     fun testHandler() {
         val request = CryptocurrencyRequest.newBuilder().build()
 
-        every { cryptocurrencyService.findAll() } returns cryptocurrencyList
+        every { cryptocurrencyService.findAll().collectList().block() } returns cryptocurrencyList
```
Comment (ajax-gnatiuk-s) [inline src/test/kotlin/com/ajax/cryptocurrency/natscontroller/NatsCryptocurrencyGetAllControllerTest.kt:53]:
```suggestion
        every { cryptocurrencyService.findAll() } returns Flux.fromIterable(cryptocurrencyList)
```

File: src/test/kotlin/com/ajax/cryptocurrency/natscontroller/NatsCryptocurrencyGetAllControllerTest.kt [hunk 6]
```diff
@@ -66,7 +66,7 @@ class NatsCryptocurrencyGetAllControllerTest {
             cryptoListFromResponse.map { it.cryptocurrencyName })
         assertEquals(cryptocurrencyList.map { it.price }, cryptoListFromResponse.map { it.price })
         assertEquals(cryptocurrencyList.map { it.createdTime }, cryptoListFromResponse.map { it.createdTime })
-        verify { cryptocurrencyService.findAll() }
+        verify { cryptocurrencyService.findAll().collectList().block() }
```
Comment (ajax-gnatiuk-s) [inline src/test/kotlin/com/ajax/cryptocurrency/natscontroller/NatsCryptocurrencyGetAllControllerTest.kt:69]:
```suggestion
        verify { cryptocurrencyService.findAll() }
```

**Review by ajax-gnatiuk-s — CHANGES_REQUESTED** [review]

File: src/test/kotlin/com/ajax/cryptocurrency/natscontroller/NatsCryptocurrencyGetAllControllerTest.kt (hunk 5 above)
Comment (Wectro20) [reply src/test/kotlin/com/ajax/cryptocurrency/natscontroller/NatsCryptocurrencyGetAllControllerTest.kt:53]: fixed

File: src/test/kotlin/com/ajax/cryptocurrency/natscontroller/NatsCryptocurrencyGetAllControllerTest.kt (hunk 6 above)
Comment (Wectro20) [reply src/test/kotlin/com/ajax/cryptocurrency/natscontroller/NatsCryptocurrencyGetAllControllerTest.kt:69]: fixed

**Review by tetty-pa — APPROVED** [review]

**Pull request merged**